### DISCUSSION TITLE
[wasm] AOT huge build time

### DISCUSSION THREAD

**Discussion by sarosigy** [unoplatform/uno#9972] [Q&A] [answered]
I ask here this first... and maybe at emscripten?

I creating a wasm head for an application, configure to AOT, the build time becomes huge, namely an hour.

I could pinpoint the main issue, that wasm-ld runs for forty-fifty minutes.
The issue is with wasm-ld that it is single threaded, at least with default parameters. While the (currently) eight CPUs are all sweating while compile and optimization / finalize phase, at linking phase only one processor is running at 100% meaning an overall 12% CPU usage.

Is there a way to run wasm-ld multithreaded?
Or should I ask at emscripten?

(( I applied some switches `<WasmShellExtraEmccFlags Include="-Wl,-lto-O0" />` with I could reduce the build time to half an hour, but the linking is still single-threaded and could be helped ))

**Comment by jeromelaban** [comment, marked as answer]
Thanks for the report. You can find documentation about this [here](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-linker-opts.html).  Indeed, the wasm linker is particularly slow (there's no way to make it run multi-threaded that I know of), and you can adjust the optimizations, or use [profiled AOT](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/runtime-execution-modes.html#profile-guided-aot), which reduces the amount of AOTed code.

Make sure to have the [IL Linker enabled](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/linker-configuration.html) as well, otherwise you'll have a lot of unused code being compiled.

**Reply by sarosigy** [reply]
OK, so all of the above and with the line `<WasmShellEmccLinkOptimizationLevel>None</WasmShellEmccLinkOptimizationLevel>` it is 16 minutes now. Not so fast, but for releasing it's fine enough.
If I'll have time I will play with profiling, and with the full optimized version, also.
